Lesson 10 hw realization

AlexandrSid opened the pull request:
Задание 10.2 ничего подобного за 5 лет не видел, и так и не увидел.
класс изменяется и только после этого грузится, но байткод соответсвует состоянию класса на момент старта программы,
доступ к компилятору из среды выполнения - пробовал найти что-то про это, нашёл всего одну публикацию от 1 Апреля 2007 года. https://javabeat.net/the-java-6-0-compiler-api/
может в другой среде выполнения это заработало бы как задумано
теория загрузки классов не даёт даже намёка на то, как это реализовать

Блин, даже запрос в гугл по такой задаче составить проблема! Самое релевантное: https://stackoverflow.com/questions/935175/convert-string-to-code

Жду комментов.

AlexandrSid committed:
Задание 10,формальное выполнение.

AlexandrSid committed:
Задание 10.2 structure

AlexandrSid committed:
Задание 10.2 ничего подобного за 5 лет не видел, и так и не увидел.
класс грузиться уже после изменения, но байткод соответсвует состоянию класса на момент старта программы, доступ к компилятору из среды выполнения - пробовал найти, не получилось.
может в другой среде выполнения это заработало бы как задумано
теория загрузки классов не даёт даже намёка на то, как это реализовать
Compiler API не стал подробно изучать, отчасти оттого, что найденный по нему материал опубликован 1 апреля https://javabeat.net/the-java-6-0-compiler-api/

Жду комментов.

efremov-itcenter commented:
Привет!
Задание с прокси - ок

С загрузкой классов придется доделать:
- разбери примеры: https://stackoverflow.com/questions/21544446/how-do-you-dynamically-compile-and-load-external-java-classes и https://blog.frankel.ch/compilation-java-code-on-the-fly/

- имплементируй свой загрузчик, который будет не только грузить файл, но и (как вариант) предварительно компилировать исходник, в помощь https://www.baeldung.com/java-classloaders

AlexandrSid committed:
Всё. это не работает. может когда-то раньше jre поволяла подменять классы, но теперь видать безопаснее стала.
Исходник меняется, компилируется, но до перезапуска программы либо не влияет на её работу, либо ломает машину.

Зато узнал про  NoClassDefFoundError
